Added export to Defold .collection files

CodeSpartan opened the pull request:
This pull request adds a plugin to Tiled, which allows exporting a map into a .collection (Defold's prefab).

Exporting into a collection allows the use of:
- Group layers (**only top-level group layers are supported, not nested ones!**)
- Multiple Tilesets per Tilemap

Upon export:
- The **Path** property of each Tileset may need to be set up manually in Defold after each export. However, Tiled will attempt to find the .tilesource file corresponding with the name of the Tileset in Tiled in your project's /tilesources/ directory. If one is found, manual adjustments won't be necessary. The way Tiled finds the root of your project is by looking for a file called "game.project" while going up the hierarchy, starting from the directory where you save the collection. Therefore, if you save your Collection outside your project, manually setting up the **Path** property will be necessary afterwards.
- If you create custom properties on your map called "x-offset" and "y-offset", these values will be used as coordinates for your top-level GameObject in the Collection.

All layers of a Tilemap will have **Z**-index property assigned with values ranging between **0** and **0.1**.
The plugin supports the use of 9999 Group Layers and 9999 Tile Layers per Group Layer.
Z-Index is assigned by combining Group Layer's order (from bottom up) with Tile Layer's order (from bottom up) in the following way: group's order is divided by 10000, layer's order is divided by 1'0000'0000, and their sum is the Z-index. So, for example, if the Group Layer's order was 15, and the Tile Layer's order was 200, the resulting Z-index would be 0.00150200. This guarantees correct superimposing of all elements in Defold.

bjorn commented on src/plugins/defoldcollection/defoldcollectionplugin.cpp:
```diff
@@ -0,0 +1,473 @@
+/*
+ * Returns a new filepath relative to the root of the Defold project if we're in one.
+ * Determines the root of the project by looking for a file called "game.project".
+ * If no such file is found by going up the heirarchy, return filename from the filepath.
```
```suggestion
 * If no such file is found by going up the hierarchy, return filename from the \a filePath.
```

bjorn commented on src/plugins/defoldcollection/defoldcollectionplugin.cpp:
```diff
@@ -0,0 +1,473 @@
+ * Determines the root of the project by looking for a file called "game.project".
+ * If no such file is found by going up the heirarchy, return filename from the filepath.
+*/
+QString DefoldCollectionPlugin::TilesetRelativePath(QString filePath)
```
Coding style: please rename to `tilesetRelativePath`.

Also please pass `QString` as `const QString &` when possible.

Finally I think this function could be a static local function instead of a member.

bjorn commented on src/plugins/defoldcollection/defoldcollectionplugin.cpp:
```diff
@@ -0,0 +1,473 @@
+{
+    QString gameproject = "/game.project";
+    QFileInfo fi(filePath);
+    QDir Qd = fi.dir();
```
Coding style: please rename to `dir`.

bjorn commented on src/plugins/defoldcollection/defoldcollectionplugin.cpp:
```diff
@@ -0,0 +1,473 @@
+    QDir Qd = fi.dir();
+
+    while (Qd.exists() && !Qd.isRoot())
+    {
```
Coding style: please put `{` on the same line (also for `if`, `for` and `else`)

bjorn commented on src/plugins/defoldcollection/defoldcollectionplugin.cpp:
```diff
@@ -0,0 +1,473 @@
+            filePath.replace(Qd.path(), "");
+            return filePath;
+        }
+        else if (!Qd.cdUp())
```
Coding style: please put `}` on the same line as the `else`.

bjorn commented on src/plugins/defoldcollection/defoldcollectionplugin.cpp:
```diff
@@ -0,0 +1,473 @@
+
+    while (Qd.exists() && !Qd.isRoot())
+    {
+        if (QFileInfo::exists(Qd.path() + gameproject))
```
I think it's nicer to remove the `/` from `gameproject` and use [`QDir::filePath`](https://doc.qt.io/qt-5/qdir.html#filePath).

bjorn commented on src/plugins/defoldcollection/defoldcollectionplugin.cpp:
```diff
@@ -0,0 +1,473 @@
+        QString components;
+
+        // create a tilemap for each tileset this map uses
+        for (auto& tileset_current : map->tilesets())
```
Coding style: please don't use underscores in variable names. Can be renamed to simply `tileset` I think. Also, please place the `&` against the variable name instead of the type.

bjorn commented on src/plugins/defoldcollection/defoldcollectionplugin.cpp:
```diff
@@ -0,0 +1,473 @@
+            map_h["tile_set"] = "/tilesources/" + tileset_current->name() + ".tilesource";
+
+            QString result = replaceTags(QLatin1String(map_t), map_h);
+            Tiled::SaveFile mapFile(tilesetFilePath);
```
Please rename to `tilesetFile` to avoid confusion. :-)

bjorn commented on src/plugins/defoldcollection/defoldcollectionplugin.cpp:
```diff
@@ -0,0 +1,473 @@
+            // write as many tilemaps as there are tilesets per group layer
+            for (auto& tileset_current : map->tilesets())
+            {
+                QString tilesetFilePath = tilesetFileDir;
```
Even though you're writing one tilemap per tileset, this is still a tile map file, so please rename to `tilemapFilePath` to avoid confusion.

bjorn commented on src/plugins/defoldcollection/defoldcollectionplugin.cpp:
```diff
@@ -0,0 +1,473 @@
+        collection_h["embedded-instances"] = embeddedInstances;
+
+        QString result = replaceTags(QLatin1String(collection_t), collection_h);
+        Tiled::SaveFile mapFile(fileName);
```
Maybe rename this to `collectionFile`?

CodeSpartan commented:
> I've left a bunch of inline comments, mostly regarding coding style.
Done, let me know if I missed anything. The only thing that I didn't change was underscores in names like "cell_t", "cell_h", because I'm assuming that this is fine? This is how it is in the original defold plugin.

> I am wondering if it still makes sense to keep the original Defold export plugin, or does it make sense to replace it with this one?
I would absolutely keep both, not everyone needs collections. Defold heavily emphasizes light-weightness and performance, and they didn't implement out-of-the-box multiple tilesets per map because each tileset adds a draw call. So I would leave both and add a warning that using multiple tilesources results in multiple drawcalls, which can introduce a source of poor performance if users mix tiles from **many** tilesets.

bjorn commented:
> Done, let me know if I missed anything. The only thing that I didn't change was underscores in names like "cell_t", "cell_h", because I'm assuming that this is fine? This is how it is in the original defold plugin.

Yeah, I realize we have precedence there. It's not my style, but I had decided to leave that be. Personally I don't like these names, because I do not know what the "t" or "h" mean (though my guesses would be "template" and "hash"). For string constants I usually use UPPERCASED_UNDERSCORED_STYLE.

> I would absolutely keep both, because not everyone needs collections. Besides, Defold heavily emphasizes light-weightness and performance, and they didn't implement out-of-the-box multiple tilesets per map because each tileset adds a draw call. So I would leave both and add a warning that using multiple tilesets results in multiple drawcalls, which can introduce a source of poor performance if users mix tiles from **many** tilesets.

Alright. You're definitely right that using multiple tilesets will result in multiple drawcalls, unless of course some technical solution is implemented to avoid that, but either will have a runtime impact that can be avoided by creating the map using only a single tileset in the first place.

But when this plugin creates the same tilemap file as the other plugin, in addition to the collection file, then somebody using Defold can always choose to ignore the collection file and use the tilemap file directly if that suits his needs better. Then we avoid maintenance and a bit of confusion on the user side about which option to choose. Alternatively, we could merge the plugins so that a single plugin exposes the two different export formats, to share some of the code.

> Regarding Travis-CI failing, it looks like it's an internal error on their side, correct?

Yes, it appears the Qt download is failing. I may need to make it more verbose somehow, but not too verbose. Right now it fails due to 10 minutes of no output while it is downloading Qt (which is quite large). But when I make the download verbose, it fails because it reaches the maximum log length...

bjorn commented on src/plugins/defoldcollection/defoldcollectionplugin.h:
```diff
@@ -0,0 +1,52 @@
+    QString mError;
+};
+
+static QString tilesetRelativePath(const QString & filePath);
```
It makes no sense to declare this function as static in the header file. The static will make its symbol local to the translation unit, so it can't be used somewhere else anyway. Just remove the declaration and move the `static` keyword to the definition.

bjorn commented on src/plugins/defoldcollection/defoldcollectionplugin.cpp:
```diff
@@ -0,0 +1,444 @@
+            return false;
+        }
+    } else {
+        // if there are group layers, each group layer is a gameobject, parented to the top gameobject, while being a parent to multiple tilemap components
```
Do you think it would be possible to remove this big distinction between "if there are group layers" and "if there are no group layers"? It seems weird to me, that if I group some of my tile layers, that it will go into the group layers branch and ignore any top-level tile layers.

CodeSpartan replied:
Good point, I actually hadn't thought that someone might want to mix the two types of structures. I'll see if I can remove the distinction.

CodeSpartan commented:
Rewrote a lot of stuff today:
- Added the possibility of having both top-level tile layers and top-level group layers
- Now tilemaps only contain layers that have cells on them
- Fixed a bug where the max amount of layers wasn't 9999, but 1000
- Removed underscores from all variable names, even ones that were "_t" and "_h"

Independently of whether you intend to keep the old export format or not, I would suggest leaving both in at least for a while, until this one is proven to be stable. I tested in all ways I could think of, but I don't think it should count as "thorough". And I don't intend on working on the world for my game until much later, so if there's any bugs, I won't catch them myself. So consider keeping both export options at least for now.

Let me know if there's anything else that needs to be corrected.

britzl commented:
Wow, this is great! @CodeSpartan could you please take a look at the requested changes so that this can get merged?

bjorn commented:
@britzl I believe @CodeSpartan has already responded to all my requested changes and I've made some tweaks myself already as well. I'm not sure why I haven't merged this yet, but I'll be sure to take another look at it this week.

stefankendall commented:
Any timeline for merging this?

bjorn commented:
@stefankendall I really need to learn how not to forget about stuff I plan to do. I tend to get driven by email notifications and my last comment generated an email I expected to get around to, but I ended up working on other things that popped up afterwards.

My next full Tiled development day is Wednesday and I have set a reminder about this in my calendar, but I'll try to get around to it earlier.

britzl commented:
@bjorn You could use something like [Pull Panda](https://github.com/marketplace/pull-panda) to get reminders on open pull requests. Its integrated into the GitHub Marketplace ecosystem and is free to use.

bjorn commented:
@britzl Hmm, it appears Pull Panda can only be installed on an organization account. Well, let's see whether using my calendar works first. :-)

britzl commented:
> @britzl Hmm, it appears Pull Panda can only be installed on an organization account. Well, let's see whether using my calendar works first. :-)

Ah, I see. Well let's hope the calendar works.

bjorn commented:
So in fact I'm working on Tiled today. I've looked over the code again and did a quick test and it seemed to work fine. Unfortunately, I didn't test the merged result so I had to push a quick follow-up fix (72ba25d354286e328565692f1d00c9b35fa7ec9c) due to a signature change.

Thanks a lot @CodeSpartan for your contribution! The plugin will be included in the next development snapshot (probably later this week). Do you think you'd have time to add a note about this new export option to the [Export Formats](https://doc.mapeditor.org/en/latest/manual/export/#defold) section of the manual?

CodeSpartan commented:
@bjorn done: https://github.com/bjorn/tiled/pull/2186
